src/query/spec.ts

```diff
@@ -127,97 +128,99 @@ export function isAggregate(specQ: SpecQuery) {
 }
 
 /**
- * @return the stack offset type for the specQuery
+ * @return The Vega-Lite `StackProperties` object that describes the stack
+ * configuration of `specQ`. Returns `null` if this is not stackable.
  */
-export function stack(specQ: SpecQuery): StackProperties & {fieldEncQ: EncodingQuery, groupByEncQ: EncodingQuery} {
-  const config = specQ.config;
-  const stacked = config ? config.stack : undefined;
-
-  // Should not have stack explicitly disabled
-  if (contains(['none', null, false], stacked)) {
+export function getVlStack(specQ: SpecQuery): StackProperties {
+  if (!hasRequiredStackProperties(specQ)) {
     return null;
   }
 
-  // Should have stackable mark
-  if (!contains(['bar', 'area'], specQ.mark)) {
-    return null;
+  const encoding = toEncoding(specQ.encodings, {schema: null, wildcardMode: 'null'});
+  const mark = specQ.mark as Mark;
+
+  return stack(mark, encoding, getStackOffset(specQ));
+}
+
+/**
+ * @return The `StackOffset` specified in `specQ`, `undefined` if none
+ * is specified.
+ */
+export function getStackOffset(specQ: SpecQuery): StackOffset {
+  for (const encQ of specQ.encodings) {
+    if (typeof encQ[Property.STACK] !== 'undefined' && !isWildcard(encQ[Property.STACK])) {
+      return encQ[Property.STACK];
+    }
   }
+  return undefined;
+}
 
-  // Should be aggregate plot
-  if (!isAggregate(specQ)) {
-    return null;
+/**
+ * @return The `Channel` in which `stack` is specified in `specQ`, or
+ * `null` if none is specified.
+ */
+export function getStackChannel(specQ: SpecQuery): Channel {
+  for (const encQ of specQ.encodings) {
+    if (typeof encQ[Property.STACK] !== 'undefined' && !isWildcard(encQ.channel)) {
+      return encQ.channel;
+    }
   }
+  return null;
+}
 
-  const stackBy = specQ.encodings.reduce((sc, encQ: EncodingQuery) => {
-    if (contains(NONPOSITION_CHANNELS, encQ.channel) && (isValueQuery(encQ) || (isFieldQuery(encQ) &&!encQ.aggregate))) {
-      sc.push({
-        channel: encQ.channel,
-        fieldDef: encQ
-      });
+/**
+ * Returns true iff the given SpecQuery has the properties defined
+ * to be a potential Stack spec.
+ * @param specQ The SpecQuery in question.
+ */
+export function hasRequiredStackProperties(specQ: SpecQuery) {
+  const requiredProps = [specQ.mark, specQ.encodings.filter(encQ => !isDisabledAutoCountQuery(encQ))];
+  for (const prop of requiredProps) {
+    if (isWildcard(prop) || containsWildcard(prop)) {
+      return false;
     }
-    return sc;
-  }, []);
+  }
+  return true;
+}
 
-  if (stackBy.length === 0) {
-    return null;
+/**
+ * Returns true iff the given property does not contain a nested wildcard.
+ * @param prop The property in question.
+ * @param opt With optional `exclude` property, which defines properties to
+ * ignore when testing for wildcards.
+ */
+function containsWildcard(prop: any, opt: {exclude?: {[key: string]: 1}} = {}) {
```

Comment:
The function and parameter name do not indicate what you describe in the comment very well. 

Consider renaming the method and use generic?